### PR TITLE
Alternative to rocksdb: Added jammdb as another lightweight option for wallet storage

### PR DESCRIPTION
**Pull request by mighty840**
# Description of change

Added support for JammDB as storage provider, in addition to RocksDB and memory storage for the wallet module.

See Jammdb [here](https://github.com/pjtatlow/jammdb)

## Motivation

Currently, rocksdb is too bulky for usage on end-devices like IoT Edge devices, android, ios Mobile phones and successive builds are not guaranteed due to breaking changes coming from up stream. Jammdb offers a simpler and light-weight option. Rocksdb can still be used and is also the default.

ETO GRUPPE wanted to have own integrated wallet and hence the fork was created for adding support with JammDB. After testing and integration with different bindings for android, ios and capacitorjs, it was found to be stable and can be used as an alternative to rocksdb.

As maintaining the fork was no longer necessary it was decided to contribute these changes back to the IOTA-SDK and also allow everyone else to use the jammdb variant. The rocksdb feature is still the default and jammdb is just another optional feature which can be used.

By removing dependency from the fork, it would be easier to always pull the latest update from IOTA-SDK rather than maintaining the fork.

## Type of change

Choose a type of change, and delete any options that are not relevant.
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Tested with wallet integration in various android and iOS apps @ ETOGRUPPE. 

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: sdk/Cargo.toml [hunk 1]
```diff
@@ -82,6 +82,7 @@ reqwest = { version = "0.11.18", default-features = false, features = [
 rocksdb = { version = "0.21.0", default-features = false, features = [
     "lz4",
 ], optional = true }
+jammdb = { version = "0.10.0", optional = true}
```
Comment (DaughterOfMars) [inline sdk/Cargo.toml]: Please order these alphabetically

File: sdk/src/wallet/core/builder.rs [hunk 2]
```diff
@@ -134,10 +134,12 @@ where
             }
         }
 
-        #[cfg(all(feature = "rocksdb", feature = "storage"))]
+        #[cfg(all(not(feature = "jammdb"), feature = "rocksdb", feature = "storage"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/builder.rs]: If rocksDb is the default, then this should be enabled regardless of that feature. I.E. if you build with `--all-features` this code should use rocksDb.

File: sdk/src/wallet/storage/adapter/jammdb.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    sync::Arc,
+};
+use tokio::sync::Mutex;
+
+/// The storage id.
+pub const STORAGE_ID: &str = "JammDB";
+
+const BUCKET_NAME: &str = "storage";
+
+impl Debug for JammdbStorageAdapter {
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/adapter/jammdb.rs:17]: Inconsistent casing

File: sdk/src/wallet/storage/adapter/jammdb.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "JammDbStorageAdapter")
+    }
+}
+
+/// Key value storage adapter.
+pub struct JammdbStorageAdapter {
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/adapter/jammdb.rs]:
```suggestion
/// Key value storage adapter.
#[derive(Clone)]
pub struct JammdbStorageAdapter {
```

File: sdk/src/wallet/storage/adapter/jammdb.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+impl JammdbStorageAdapter {
+    /// Initialises the storage adapter.
+    pub fn new(path: impl AsRef<Path>) -> crate::wallet::Result<Self> {
+        let mut db_path = PathBuf::from("./sdk-wallet.db");
+        let dir_path = path.as_ref().to_string_lossy().to_string();
+        let mut temp_path = PathBuf::from(dir_path);
+        if path.as_ref().is_dir() {
+            temp_path.push(db_path);
+        }
+        db_path = temp_path;
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/adapter/jammdb.rs]:
```suggestion
        let mut db_path = path.as_ref().to_owned();
        if db_path.is_dir() {
            db_path.push("sdk-wallet.db");
        }
```

File: sdk/src/wallet/storage/adapter/mod.rs [hunk 6]
```diff
@@ -4,9 +4,16 @@
 pub mod memory;
 /// RocksDB storage adapter.
 #[cfg(feature = "rocksdb")]
+#[cfg(not(feature = "jammdb"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/adapter/mod.rs]:
```suggestion
```

File: sdk/src/wallet/storage/constants.rs [hunk 7]
```diff
@@ -4,14 +4,20 @@
...
+#[cfg(feature = "jammdb")]
+pub(crate) const JAMMDB_FOLDERNAME: &str = "walletdb";
+
 /// The default RocksDB storage path.
 #[cfg(feature = "rocksdb")]
 pub(crate) const ROCKSDB_FOLDERNAME: &str = "walletdb";
 
 pub const fn default_storage_path() -> &'static str {
-    #[cfg(feature = "rocksdb")]
+    #[cfg(all(not(feature = "jammdb"), feature = "rocksdb"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/constants.rs]: maybe use [cfg_if](https://docs.rs/cfg-if/latest/cfg_if/) to clean this up.

File: sdk/src/wallet/storage/kind.rs [hunk 8]
```diff
@@ -9,7 +9,12 @@ use serde::{Deserialize, Serialize};
 pub enum StorageKind {
     /// RocksDB storage.
     #[cfg(feature = "rocksdb")]
+    #[cfg(not(feature = "jammdb"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/kind.rs]:
```suggestion
```

File: sdk/src/wallet/storage/adapter/jammdb.rs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    type Error = crate::wallet::Error;
+
+    async fn get_bytes(&self, key: &str) -> Result<Option<Vec<u8>>, Self::Error> {
+        let db = self.db.lock().await;
+        let tx = db.tx(false)?;
+        let bucket = tx.get_bucket(BUCKET_NAME)?;
+        match bucket.get(key) {
+            Some(r) => Ok(Some(r.kv().value().into())),
+            None => Ok(None),
+        }
```
Comment (Thoralf-M) [inline sdk/src/wallet/storage/adapter/jammdb.rs]: I think this can be just
```suggestion
        Ok(bucket.get(key).map(|r| r.kv().value().into()))
```

File: sdk/src/wallet/storage/constants.rs [hunk 10]
```diff
@@ -4,14 +4,20 @@
 /// The default storage path.
 pub const DEFAULT_STORAGE_PATH: &str = "./storage";
 
+/// The default RocksDB storage path.
```
Comment (Thoralf-M) [inline sdk/src/wallet/storage/constants.rs]:
```suggestion
/// The default jammdb storage path.
```

Comment (mighty840) [reply sdk/src/wallet/storage/constants.rs]: Fixed in new commits

File: sdk/src/wallet/storage/adapter/jammdb.rs (hunk 9 above)
Comment (mighty840) [reply sdk/src/wallet/storage/adapter/jammdb.rs]: Thanks for the suggestion. Patched in new commits

File: sdk/src/wallet/core/builder.rs (hunk 2 above)
Comment (mighty840) [reply sdk/src/wallet/core/builder.rs]: Have removed unneccessary feature flags for rocskdb now from all instances that I could find.

File: sdk/src/wallet/storage/adapter/jammdb.rs (hunk 3 above)
Comment (mighty840) [reply sdk/src/wallet/storage/adapter/jammdb.rs:17]: Did not understand which part had case problems. BUCKET_NAME is consistent with const definitions and JammdbStorageAdapter is consistent with RocksdbStorageAdapter

File: sdk/src/wallet/storage/adapter/jammdb.rs (hunk 4 above)
Comment (mighty840) [reply sdk/src/wallet/storage/adapter/jammdb.rs]: Added trait

File: sdk/Cargo.toml (hunk 1 above)
Comment (mighty840) [reply sdk/Cargo.toml]: done

File: sdk/Cargo.toml [hunk 11]
```diff
@@ -67,6 +67,7 @@ heck = { version = "0.4.1", default-features = false, optional = true }
 instant = { version = "0.1.12", default-features = false, optional = true }
 iota-ledger-nano = { version = "1.0.0-alpha.4", default-features = false, optional = true }
 iota_stronghold = { version = "2.0.0", default-features = false, optional = true }
+jammdb = { version = "0.10.0", optional = true }
```
Comment (Thoralf-M) [inline sdk/Cargo.toml]: We add this to all crates so we don't get features we don't actually need, for consistency it should also be here even if there are not features at all (at least currently)
```suggestion
jammdb = { version = "0.10.0", default-features = false, optional = true }
```

File: sdk/src/wallet/storage/adapter/jammdb.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/wallet/storage/adapter/jammdb.rs:17]: Below you use `JammDbStorageAdapter` for the string

File: sdk/src/wallet/storage/constants.rs [hunk 12]
```diff
@@ -4,16 +4,16 @@
 /// The default storage path.
 pub const DEFAULT_STORAGE_PATH: &str = "./storage";
 
-/// The default RocksDB storage path.
-#[cfg(feature = "jammdb")]
+/// The default jammDB storage path.
+#[cfg(all(not(feature = "rocksdb"), feature = "jammdb"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/constants.rs]: For completely independent items like this, they should rely only on the `jammdb` feature and be additive.

File: sdk/src/wallet/storage/constants.rs [hunk 13]
```diff
@@ -4,16 +4,16 @@
...
 pub(crate) const JAMMDB_FOLDERNAME: &str = "walletdb";
 
 /// The default RocksDB storage path.
 #[cfg(feature = "rocksdb")]
 pub(crate) const ROCKSDB_FOLDERNAME: &str = "walletdb";
 
 pub const fn default_storage_path() -> &'static str {
-    #[cfg(all(not(feature = "jammdb"), feature = "rocksdb"))]
+    #[cfg(feature = "rocksdb")]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/constants.rs]: This could benefit from `cfg_if` too

Comment (DaughterOfMars) [reply sdk/src/wallet/storage/constants.rs]: Just noticed I commented the same thing twice here 😆  that's now important it is obviously

Comment (mighty840) [reply sdk/src/wallet/storage/constants.rs]: Unable to understand your comment and what exactly is expected here.

Comment (mighty840) [reply sdk/src/wallet/storage/constants.rs]: I have pushed some modifications. You can review them now.

File: sdk/Cargo.toml [hunk 14]
```diff
@@ -21,6 +21,7 @@ rustdoc-args = ["--cfg", "docsrs"]
 bech32 = { version = "0.9.1", default-features = false }
 bitflags = { version = "2.3.3", default-features = false }
 bytemuck = { version = "1.13.1", default-features = false }
+cfg-if={version= "1.0.0", default-features = false}
```
Comment (DaughterOfMars) [inline sdk/Cargo.toml]: spacing

File: sdk/src/wallet/storage/constants.rs [hunk 15]
```diff
@@ -4,15 +4,26 @@
...
 pub(crate) const ROCKSDB_FOLDERNAME: &str = "walletdb";
 
 pub const fn default_storage_path() -> &'static str {
-    #[cfg(feature = "rocksdb")]
-    return ROCKSDB_FOLDERNAME;
-    #[cfg(not(feature = "rocksdb"))]
-    DEFAULT_STORAGE_PATH
+    cfg_if::cfg_if!{
+        if #[cfg(feature = "rocksdb")]{
+            return ROCKSDB_FOLDERNAME;
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/constants.rs]: Do these need to be returns?

Comment (mighty840) [reply sdk/src/wallet/storage/constants.rs]: They were returns before. I just added cfg_if enclosures to make it look pretty.

Comment (DaughterOfMars) [reply sdk/src/wallet/storage/constants.rs]: Yes, but they don't need to be anymore.

Comment (mighty840) [reply sdk/src/wallet/storage/constants.rs]: removed them now

File: sdk/src/wallet/core/builder.rs [hunk 16]
```diff
@@ -137,7 +137,9 @@ where
         #[cfg(all(feature = "rocksdb", feature = "storage"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/builder.rs]: Use `cfg-if` here too, and waterfall through the features.

File: sdk/src/wallet/core/builder.rs [hunk 17]
```diff
@@ -15,7 +15,7 @@ use tokio::sync::RwLock;
 use super::operations::storage::SaveLoadWallet;
 #[cfg(feature = "events")]
 use crate::wallet::events::EventEmitter;
-#[cfg(all(feature = "storage", not(feature = "rocksdb")))]
+#[cfg(all(feature = "storage", not(feature = "rocksdb"), not(feature = "jammdb")))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/builder.rs]: Should probably remove this complicated import in favor of fully qualified path below with the cfg-if

File: sdk/src/wallet/storage/adapter/mod.rs [hunk 18]
```diff
@@ -7,6 +7,12 @@ pub mod memory;
 #[cfg_attr(docsrs, doc(cfg(feature = "rocksdb")))]
 pub mod rocksdb;
 
+/// JammDB storage adapter.
+#[cfg(feature = "jammdb")]
+#[cfg(not(feature = "rocksdb"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/adapter/mod.rs]:
```suggestion

```

File: sdk/src/wallet/storage/kind.rs [hunk 19]
```diff
@@ -10,6 +10,10 @@ pub enum StorageKind {
     /// RocksDB storage.
     #[cfg(feature = "rocksdb")]
     Rocksdb,
+    /// JammDB storage
+    #[cfg(feature = "jammdb")]
+    #[cfg(not(feature = "rocksdb"))]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/kind.rs]:
```suggestion
```

File: sdk/src/wallet/storage/kind.rs [hunk 20]
```diff
@@ -21,9 +25,12 @@ impl Default for StorageKind {
     fn default() -> Self {
         #[cfg(feature = "rocksdb")]
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/kind.rs]: cfg-if here

File: sdk/src/wallet/storage/kind.rs [hunk 21]
```diff
@@ -19,11 +22,25 @@ pub enum StorageKind {
...
+            if #[cfg(feature="rocksdb")]{
+                return Self::Rocksdb;
+            } else if #[cfg(feature="jammdb")] {
+                return Self::Jammdb;
+            }
+            else if #[cfg(target_family="wasm")]{
+                return Self::Wasm;
+            }else{
+                return Self::Memory;
+            }
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/kind.rs]:
```suggestion
            if #[cfg(feature = "rocksdb")] {
                Self::Rocksdb
            } else if #[cfg(feature = "jammdb")] {
                Self::Jammdb
            } else if #[cfg(target_family = "wasm")] {
                Self::Wasm
            } else {
                Self::Memory
            }
```

File: sdk/src/wallet/core/builder.rs [hunk 22]
```diff
@@ -134,11 +132,19 @@ where
...
+                let storage =
+                crate::wallet::storage::adapter::rocksdb::RocksdbStorageAdapter::new(storage_options.path.clone())?;
+            }
+            else if #[cfg(feature = "jammdb")]{
+                let storage = crate::wallet::storage::adapter::jammdb::JammdbStorageAdapter::new(storage_options.path.clone())?;
+            }
+            else {
+                let storage = crate::wallet::storage::adapter::memory::Memory::default();
+            }
+        );
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/builder.rs]:
```suggestion
        let storage = cfg_if::cfg_if!(
            if #[cfg(feature = "rocksdb")] {
                crate::wallet::storage::adapter::rocksdb::RocksdbStorageAdapter::new(storage_options.path.clone())?
            } else if #[cfg(feature = "jammdb")] {
                crate::wallet::storage::adapter::jammdb::JammdbStorageAdapter::new(storage_options.path.clone())?
            } else {
                crate::wallet::storage::adapter::memory::Memory::default()
            }
        );
```

File: sdk/src/wallet/storage/constants.rs [hunk 23]
```diff
@@ -4,15 +4,26 @@
...
+        if #[cfg(feature = "rocksdb")]{
+            ROCKSDB_FOLDERNAME
+        }
+        else if #[cfg(feature = "jammdb")] {
+            JAMMDB_FOLDERNAME
+        }
+        else {
+            DEFAULT_STORAGE_PATH
+        }
+    }
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/constants.rs]:
```suggestion
    cfg_if::cfg_if! {
        if #[cfg(feature = "rocksdb")] {
            ROCKSDB_FOLDERNAME
        } else if #[cfg(feature = "jammdb")] {
            JAMMDB_FOLDERNAME
        } else {
            DEFAULT_STORAGE_PATH
        }
    }
```

File: sdk/src/wallet/storage/kind.rs [hunk 24]
```diff
@@ -19,11 +22,25 @@ pub enum StorageKind {
...
+        );
+        // #[cfg(feature = "rocksdb")]
+        // return #[cfg(feature = "jammdb")]
+        // #[cfg(not(feature = "rocksdb"))]
+        // return Self::Jammdb;
+        // #[cfg(target_family = "wasm")]
+        // return Self::Wasm;
+        // #[cfg(not(any(feature = "rocksdb", target_family = "wasm", feature = "jammdb")))]
+        // Self::Memory
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/kind.rs]:
```suggestion
```

**Comment by mighty840** [comment]
The suggested changes have introduced issues related to cfg_if while performing cargo check. Do I need to revert back the changes? I tested it with rustc 1.71.0 (8ede3aae2 2023-07-12)

**Comment by DaughterOfMars** [comment]
What is the issue?

**Comment by mighty840** [comment]
cargo check fails on my machine for the given changes. See log here:

```
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /home/sparnerkar/workspace/iota-sdk/bindings/nodejs/Cargo.toml
workspace: /home/sparnerkar/workspace/iota-sdk/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /home/sparnerkar/workspace/iota-sdk/bindings/nodejs-old/Cargo.toml
workspace: /home/sparnerkar/workspace/iota-sdk/Cargo.toml
    Checking iota-sdk v1.0.2 (/home/sparnerkar/workspace/iota-sdk/sdk)
error: macro expansion ignores token `$crate` and any following
   --> /home/sparnerkar/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cfg-if-1.0.0/src/lib.rs:79:9
    |
79  |           $crate::cfg_if! { @__items ($($not,)* $($m,)*) ; $($rest)* }
    |           ^^^^^^
    |
   ::: sdk/src/wallet/core/builder.rs:136:23
    |
136 |           let storage = cfg_if::cfg_if!(
    |  _______________________-
137 | |             if #[cfg(feature = "rocksdb")] {
138 | |                 crate::wallet::storage::adapter::rocksdb::RocksdbStorageAdapter::new(storage_options.path.clone())?
139 | |             } else if #[cfg(feature = "jammdb")] {
...   |
143 | |             }
144 | |         );
    | |_________- caused by the macro expansion here
    |
    = note: the usage of `cfg_if!` is likely invalid in expression context
help: you might be missing a semicolon here
    |
46  |         };
    |          +

error[E0658]: attributes on expressions are experimental
   --> sdk/src/wallet/core/builder.rs:136:23
    |
136 |           let storage = cfg_if::cfg_if!(
    |  _______________________^
137 | |             if #[cfg(feature = "rocksdb")] {
138 | |                 crate::wallet::storage::adapter::rocksdb::RocksdbStorageAdapter::new(storage_options.path.clone())?
139 | |             } else if #[cfg(feature = "jammdb")] {
...   |
143 | |             }
144 | |         );
    | |_________^
    |
    = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
    = note: this error originates in the macro `$crate::cfg_if` which comes from the expansion of the macro `cfg_if::cfg_if` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> sdk/src/wallet/storage/kind.rs:24:21
   |
24 |     fn default() -> Self {
   |        -------      ^^^^ expected `StorageKind`, found `()`
   |        |
   |        implicitly returns `()` as its body has no tail or `return` expression
...
27 |                 Self::Rocksdb
   |                             - help: remove this semicolon to return this value

Some errors have detailed explanations: E0308, E0658.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `iota-sdk` (lib) due to 3 previous errors
```

File: sdk/src/wallet/core/builder.rs [hunk 25]
```diff
@@ -134,11 +132,16 @@ where
...
+        #[cfg(feature = "storage")]
+        let storage = cfg_if::cfg_if!(
+            if #[cfg(feature = "rocksdb")] {
+                crate::wallet::storage::adapter::rocksdb::RocksdbStorageAdapter::new(storage_options.path.clone())?
+            } else if #[cfg(feature = "jammdb")] {
+                crate::wallet::storage::adapter::jammdb::JammdbStorageAdapter::new(storage_options.path.clone())?
+            } else {
+                crate::wallet::storage::adapter::memory::Memory::default()
+            }
+        );
```
Comment (DaughterOfMars) [inline sdk/src/wallet/core/builder.rs]:
```suggestion
        cfg_if::cfg_if!(
            if #[cfg(feature = "rocksdb")] {
                let storage = crate::wallet::storage::adapter::rocksdb::RocksdbStorageAdapter::new(storage_options.path.clone())?;
            } else if #[cfg(feature = "jammdb")] {
                let storage = crate::wallet::storage::adapter::jammdb::JammdbStorageAdapter::new(storage_options.path.clone())?;
            } else {
                let storage = crate::wallet::storage::adapter::memory::Memory::default();
            }
        )
```

guess you can't use cfg-if as an expression 🤷‍♀️

**Comment by DaughterOfMars** [comment]
Please resolve the conflicts and then we should be good to merge

**Comment by mighty840** [comment]
I think I need to add return statement in sdk/src/wallet/storage/kind.rs default() method. Cargo check still throws error. This might be because cfg_if is adding some semi-colons. See log below:

warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /home/sparnerkar/workspace/iota-sdk/bindings/nodejs/Cargo.toml
workspace: /home/sparnerkar/workspace/iota-sdk/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /home/sparnerkar/workspace/iota-sdk/bindings/nodejs-old/Cargo.toml
workspace: /home/sparnerkar/workspace/iota-sdk/Cargo.toml
    Checking iota-sdk v1.0.2 (/home/sparnerkar/workspace/iota-sdk/sdk)
error[E0308]: mismatched types
  --> sdk/src/wallet/storage/kind.rs:24:21
   |
24 |     fn default() -> Self {
   |        -------      ^^^^ expected `StorageKind`, found `()`
   |        |
   |        implicitly returns `()` as its body has no tail or `return` expression
...
27 |                 Self::Rocksdb
   |                             - help: remove this semicolon to return this value

For more information about this error, try `rustc --explain E0308`.
error: could not compile `iota-sdk` (lib) due to previous error

File: sdk/src/wallet/storage/kind.rs [hunk 26]
```diff
@@ -19,11 +22,16 @@ pub enum StorageKind {
...
+            if #[cfg(feature = "rocksdb")] {
+                Self::Rocksdb
+            } else if #[cfg(feature = "jammdb")] {
+                Self::Jammdb
+            } else if #[cfg(target_family = "wasm")] {
+                Self::Wasm
+            } else {
+                Self::Memory
+            }
+        );
```
Comment (DaughterOfMars) [inline sdk/src/wallet/storage/kind.rs]:
```suggestion
        )
```

**Comment by DaughterOfMars** [comment]
I fixed that issue. Please resolve conflicts

**Comment by thibault-martinez** [comment]
Also, can you please change the base branch to 1.1? We'd probably release that together with other updates.

File: sdk/src/wallet/storage/kind.rs (hunk 26 above)
Comment (mighty840) [reply sdk/src/wallet/storage/kind.rs]: The issue still persists. I dont know if we can get away by avoiding those return statements

Comment (mighty840) [reply sdk/src/wallet/storage/kind.rs]: Suggested change...

```diff
diff --git a/sdk/src/wallet/storage/kind.rs b/sdk/src/wallet/storage/kind.rs
index 7c883b0b6..8b753e45e 100644
--- a/sdk/src/wallet/storage/kind.rs
+++ b/sdk/src/wallet/storage/kind.rs
@@ -24,14 +24,14 @@ impl Default for StorageKind {
     fn default() -> Self {
         cfg_if::cfg_if!(
             if #[cfg(feature = "rocksdb")] {
-                Self::Rocksdb
+                return Self::Rocksdb;
             } else if #[cfg(feature = "jammdb")] {
-                Self::Jammdb
+                return Self::Jammdb;
             } else if #[cfg(target_family = "wasm")] {
-                Self::Wasm
+                return Self::Wasm;
             } else {
-                Self::Memory
+                return Self::Memory;
             }
-        )
+        );
     }
 }
```

**Comment by DaughterOfMars** [comment]
I don't think the merge went okay. Also, regarding the `cfg-if` macros, I think the problem is actually that you used parenthesis `()` instead of brackets `{}`.

**Comment by mighty840** [comment]
ahh yes. I understood my error. Fixing it now. It works after replacing () with {}. cargo check also works! Thanks for the input.

**Comment by DaughterOfMars** [comment]
The merge is seriously eff'd. You may want to start over if you don't feel that you can fix it

**Comment by mighty840** [comment]
I dont understand what remains now to be fixed.

**Comment by mighty840** [comment]
I have resolved the newer conflicts. Its very hard to understand the cryptic and unregulated comments as well as hard to follow two lines of action coming in from different devs. If you feel like merging this in the up stream, let us know, otherwise we will keep maintaining the fork, but I think there is no point for us now to waste our time beautifying and "eff'ing" your code in this PR.

Also, would like to know where to report code of conduct violations.

**Comment by thibault-martinez** [comment]
Hi @mighty840,

> Its very hard to understand the cryptic and unregulated comments

I've read everything again and honestly don't see anything cryptic at all. Can you give me an example?
Also please be sure you can always ask for clarification and we will happily answer.

> hard to follow two lines of action coming in from different devs

We require 2 approvals from the team (and even sometimes 4) to merge PRs. This is a very standard practice so you will always have different devs giving their opinion. Please have a look at ALL our other PRs and you will see that this one hasn't been treated any different. 

> If you feel like merging this in the up stream

We are not sure yet, we need to spend some time investigating that library before we take any decision.

> waste our time beautifying and "eff'ing" your code in this PR.

I am sorry you feel this way but we are not going to reduce our quality standards. Again, this PR is treated as any other PR, internal or external.

> Also, would like to know where to report code of conduct violations.

Feel free to send a mail to thibault@iota.org.

**Review by kwek20 — APPROVED** [review]

**Comment by thibault-martinez** [comment]
Hi @mighty840, sorry for the delay, we have been super busy with stardust on mainnet and work on 2.0. We have taken the time to think about this and for now it seems that the best solution is for you to continue maintaining your fork. With 2.0, we have the opportunity to do breaking changes and completely revisit our storage layer so we will definitely take your input regarding IoT devices in consideration. Thanks for your understanding.

**Comment by mighty840** [comment]
@thibault-martinez works for us. Keep us posted if you need to merge it later. We will try to keep our fork patched as far as we can.

**Comment by thibault-martinez** [comment]
> @thibault-martinez works for us. Keep us posted if you need to merge it later. We will try to keep our fork patched as far as we can.

Thank you for your understanding. I will close the PR until we start working on the database layer for 2.0, is that fine with you? Also, develop won't change too much now so it should be fairly easy to maintain your fork.

**Comment by mighty840** [comment]
Yes. I am closing the PR now from my side.
